base/tps/src/org/dogtagpki/server/tps/TPSTokendb.java

```diff
@@ -243,25 +243,76 @@ public void tdbAddCertificatesForCUID(String cuid, ArrayList<TPSCertRecord> cert
 
         logger.debug(method + " found token " + cuid);
         logger.debug(method + " number of certs to update:" + certs.size());
+
+	// Keep track of which certs made it to the database and which didn't,
+        // in case of failure
+         class CnIssuerPair {
+             public final String cn;
+             public final String issuer;
+
+             public CnIssuerPair(String _cn, String _issuer) {
+                 cn = _cn;
+                 issuer = _issuer;
+             }
+
+             public String toString() {
+                 return "(cn=" + cn + ", issuerCn=" + issuer + ")";
+             }
+         }
+
+         ArrayList<CnIssuerPair> cnIssuerPairsRemaining = new ArrayList<CnIssuerPair>(certs.size());
+         for(TPSCertRecord cert : certs) {
+             String cn = cert.getId();
+             String issuerCn = cert.getIssuedBy();
+             cnIssuerPairsRemaining.add(new CnIssuerPair(cn, issuerCn));
+         }
+
+        boolean testAddCertsFailure = false;
+	//Contrive a very difficult to reproduce testing scenario
+	CMSEngine engine = CMS.getCMSEngine();
+	try {
+	    EngineConfig configStore = engine.getConfig();
+            // get conn ID
+            String config = "op.enroll." + "testAddCertsToDBFailure";
+            testAddCertsFailure = configStore.getBoolean(config,false);
+        } catch (Exception e) {
+           testAddCertsFailure = false;
+        }
+
         try {
+	    int count = 0;
             for (TPSCertRecord cert : certs) {
                 try {
                     if (!isCertOnToken(cert, cuid)) {
                         logger.debug(method + " adding cert with serial: " + cert.getSerialNumber());
+			// After at least one cert is added correctly, perform the test of a failure
+                        // if so configured.
+                        if(count > 0 && testAddCertsFailure == true) {
+                            throw new Exception("Failed to add certificate to token db, as part of a test of failure condition.");
```

Comment:
Regardless of the fact that it's just a test. please try to adhere to the practice of adding the content of "method" to the log message to assist admins in pinpointing the location where the message comes from.